project/MimaExcludes.scala

```diff
@@ -104,6 +104,10 @@ object MimaExcludes {
     ProblemFilters.exclude[DirectMissingMethodProblem]("org.apache.spark.shuffle.api.SingleSpillShuffleMapOutputWriter.transferMapSpillFile"),
     ProblemFilters.exclude[ReversedMissingMethodProblem]("org.apache.spark.shuffle.api.SingleSpillShuffleMapOutputWriter.transferMapSpillFile"),
     ProblemFilters.exclude[ReversedMissingMethodProblem]("org.apache.spark.shuffle.api.ShuffleMapOutputWriter.commitAllPartitions")
+
+    // [SPARK-36419][CORE] Optionally move final aggregation in RDD.treeAggregate to executor
+    ProblemFilters.exclude[ReversedMissingMethodProblem]("org.apache.spark.api.java.JavaRDDLike.treeAggregate")
```

Comment:
@venkata91 and I tried doing this and ran into a couple issues while trying to write the overloaded function. Seems like the issue is scala while resolving overloaded methods, it looks at the first argument and if the type is same on both the methods then that becomes ambiguous. Here is the version of the overloaded method we were trying with (along with several other variations of this :) ): 

```

  def treeAggregate[U: ClassTag](zeroValue: U)(
    seqOp: (U, T) => U,
    combOp: (U, U) => U,
    depth: Int = 2): U = withScope {
    treeAggregate(zeroValue)(seqOp, combOp, depth, false)
  }
```
any suggestions? @srowen

Reply:
You need to declare two methods, and not use optional args

Reply:
I think the issue is because of `zeroValue:U` is only inspected by the compiler to resolve the overloaded methods. Due to that the compiler throws `ambiguous reference to overloaded definition`. Something like this cannot be resolved - `val func = rdd.treeAggregate(0L) _` properly with the newly added overloaded method. @srowen What do you think?

Reply:
Hm. Two separate methods won't work here b/c of the currying (or whatever it's called)? the thing is we need to not break API compatibility over this. If there's no apparent way around it ... hm. We could make a differently named new method which is kind of ugly, or fall back to a config param?

Reply:
since we already decided that using parameters was the better option compared to configs, let's make a new function?

Reply:
or we can revert back to the older version with configs.

Reply:
Yes it doesn't work because of the currying here I think. `config` seems like an easy workaround rather than creating a new method.

Reply:
@srowen i dont think we break api compatibility on the scala side because we use a default param anyways, so the user would not have to adapt to the new signature. So i am proposing that we keep the scala file in the mima excludes and not worry about the binary incompatibility for that.

On the java side, we will be breaking api compatibility, and we can add a method there that is overloaded. This will allow us to not break compatibility. There's already 2 `treeAggregate` methods in the javaRdd, so roughly speaking we will just add a 3rd one. Something like this: 
```
  /**
   * Aggregates the elements of this RDD in a multi-level tree pattern.
   *
   * @param depth suggested depth of the tree
   * @see [[org.apache.spark.api.java.JavaRDDLike#aggregate]]
   */
  def treeAggregate[U](
    zeroValue: U,
    seqOp: JFunction2[U, T, U],
    combOp: JFunction2[U, U, U],
    depth: Int): U = {
    rdd.treeAggregate(zeroValue)(seqOp, combOp, depth)(fakeClassTag[U])
  }

  /**
   * `org.apache.spark.api.java.JavaRDDLike.treeAggregate` with suggested depth 2.
   */
  def treeAggregate[U](
    zeroValue: U,
    seqOp: JFunction2[U, T, U],
    combOp: JFunction2[U, U, U]): U = {
    treeAggregate(zeroValue, seqOp, combOp, 2)
  }

  /**
   * Aggregates the elements of this RDD in a multi-level tree pattern and provides the option of doing 
   * the final aggregation on an executor
   *
   * @param depth suggested depth of the tree
   * @see [[org.apache.spark.api.java.JavaRDDLike#aggregate]]
   */
  def treeAggregate[U](
    zeroValue: U,
    seqOp: JFunction2[U, T, U],
    combOp: JFunction2[U, U, U],
    depth: Int,
    finalAggregateOnExecutor: Boolean): U = {
    rdd.treeAggregate(zeroValue)(seqOp, combOp, depth, finalAggregateOnExecutor)(fakeClassTag[U])
  }
```
Then we can remove the mima exludes for the javaRdd class. what do you think?

Reply:
No, it may be source compatible, but not binary compatible in Scala, and this isn't worth breaking an API over.
Yes you're right there is a similar consideration for Java-specific methods.

Reply:
okay, so i think the best option we have now is to revert back to the earlier version that used a config. not sure if making a new method is worth it.

Reply:
^^ do we have consensus on that? @mridulm @venkata91 @srowen

Reply:
Given the situation, +1 on adding this change behind a config unless there is a strong reason not to do so.